### ISSUE TITLE
Allow dedupe statuses when scheduling new orchestration instance

### ISSUE THREAD

**Issue by sagimar** [microsoft/durabletask-dotnet#183]
Hi, we wish to achieve idempotence when calling ScheduleNewOrchestrationInstanceAsync with the same orchestration id, so if the orchestration id is already existed (in pending, running or completion state)- the orchestration won't be re-triggered.

In the current implementation, the orchestration is actually being triggered twice, even if the orchestration is in Pending/Running state, and the documentation has no real promises in this matter:

> "If an orchestration with the specified instance ID already exists and is in a non-terminal state (Pending, Running, etc.), then this operation may fail silently. However, if an orchestration instance with this ID already exists in a terminal state (Completed, Terminated, Failed, etc.) then the instance may be recreated automatically, depending on the configuration of the backend instance store."

Suggested solution:
To schedule a new orchestration instance, one should call:
```
    public override async Task<string> ScheduleNewOrchestrationInstanceAsync(
        TaskName orchestratorName,
        object? input = null,
        StartOrchestrationOptions? options = null,
        CancellationToken cancellation = default)
```

The implementation of this method is in [ShimDurableTaskClient.cs](https://github.com/microsoft/durabletask-dotnet/blob/e2f329258d1e02a762911bd89274a9a584b6ec04/src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs#L129) and in the end of the methods it calls:
`await this.Client.CreateTaskOrchestrationAsync(message);`

This method is from azure/durabletask, and it has another overload, with the following signature:
`public Task CreateTaskOrchestrationAsync(TaskMessage creationMessage, OrchestrationStatus[]? dedupeStatuses)`
When one sending the dedupeStatuses to this method, if an orchestration is existed in one of the given statuses, durabletask throws InvalidOperationException, that could be caught to handle idempotence.

Then, in my opinion, allowing receiving OrchestrationStatus[]? dedupeStatuses in ScheduleNewOrchestrationInstanceAsync() should help to handle the idempotence (or any other solution you have in mind will be appreciated).

Thanks,
Sagi

**Comment by davidmrdavid** [comment]
Hi @sagimar: thanks for reaching out.
I think I need more context here: when you say that "In the current implementation, the orchestration is actually being triggered twice", do you mean that you're seeing the orchestration being triggered twice one after the other, or do you mean in parallel? Just trying to get clarity around the current behavior you're seeing. Any traces that you may share will help me understand your requirements.

**Comment by sagimar** [comment]
Hi, thanks for the response!
I'll further explain the issue:
I'm making the following calls (scheduling the same orchestration id twice):

```
public async Task ScheduleOrchestrations()
{
    await client.ScheduleNewOrchestrationInstanceAsync("orchestrationId", input, options, cancellationToken).ConfigureAwait(false);

    await client.ScheduleNewOrchestrationInstanceAsync("orchestrationId", input, options, cancellationToken).ConfigureAwait(false);
}
```

The expected behavior: Creation of one orchestration from the first call, and no-op in the second call (idempotence).

The behavior- this is the logs I see from my service:
1. Fetched instance status - RuntimeStatus: NotFound
2. Sending [ExecutionStarted] message, SequenceNumber: 1
3. Updated instances table and set the runtime status to 'Pending'
4. Fetched instance status - RuntimeStatus: Pending
5. Sending [ExecutionStarted] message, SequenceNumber: 2
6. Updated instance table and set the runtime status to 'Pending'

Logs 1-3 are from the first call and 4-6 are from the seconds call. Same happens also if the orchestration was in 'Running' state.

I would expect that after log number (4)- no-op will occur. This could be achieved using dedupeStatuses functionality from azure/durabletask (it will throw an exception instead of sending ExecutionStarted event and updating the instance table).

Thanks!

**Comment by davidmrdavid** [comment]
Just to confirm, you're saying that you sometimes see an orchestrator already in the `Running` state get set back to `Pending` when you call `await client.ScheduleNewOrchestrationInstanceAsync`?. That seems like a bug to me. The case where `Pending` is set again to `Pending` seems safe enough, but the former case is more problematic.

**Comment by sagimar** [comment]
Yes- this is absolutely what happens (even not sometimes, using storage).
But more than that, I would also want to have the ability to control the idempotence level, so for example also if the state is 'Completed' (also 'Pending' in my opinion)- it won't re-schedule the orchestration.

**Comment by violetVo** [comment]
@davidmrdavid - any update on this? this is a blocking bug since it breaks the idempotency. As far as I understand from @sagimar comment above, this is a very easy fix on your side (just to expose the extra parameter and call the the method's overload).
Is there an ETA for the fix?

**Comment by davidmrdavid** [comment]
FYI @jviau - would be good if you could take a look at this one

**Comment by violetVo** [comment]
@jviau  - Any update on this fix? this is a critical bug, and we are blocked on this for a very long time.

**Comment by epDugas** [comment]
Came across this, and looking for verification/validation that this was an actual bug that allowed a singleton to be in a running state, and be overwritten by a subsequent request, before the dedupe statuses where added to the start options?
